### PR TITLE
[C++/PyTorch] Add alibi_slopes support

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR adds `alibi_slopes` support to `FlashAttention`, `FusedAttention` and `UnfusedDotProductAttention`. It allows users to specify an `alibi_slopes` tensor with custom values, which can be different from values calculated using the formula in the [original paper](https://arxiv.org/abs/2108.12409).

`FlashAttention`:  supports shapes `[num_heads]` and `[batch_size, num_heads]` for `alibi_slopes`. To utilize this functionality, `core_attention_bias_type` must be `alibi`.

`FusedAttention`: supports only shape `[num_heads]`. This is implemented through the `post_scale_bias` code path in the `FusedAttention` module (both `F16_arbitrary_seqlen` or `F16_max512_seqlen` backends).

`UnfusedDotProductAttention`: supports both `[num_heads]` and `[batch_size, num_heads]` shapes.

### CODE REVIEW

File: tests/pytorch/fused_attn/test_fused_attn.py [hunk 1]
```diff
@@ -381,6 +393,21 @@ def test_dpa_sliding_window(dtype, model_configs, model):
     """Test DotProductAttention module with sliding window attention"""
     test_dot_product_attention(dtype, model_configs, model, False, True, None, True)
 
+model_configs_alibi_slopes = {
+    #     test:             b,  h, hg,   d,   sq,  skv,   p,      mask,    bias, alibi_type
+    "alibi_1_0": ModelConfig(4, 16, 16,  64,  128,  128, 0.0, "causal", "alibi", alibi_type="vanilla"),
+    "alibi_1_1": ModelConfig(2, 16, 16,  64,  128,  256, 0.0, "causal", "alibi", alibi_type="vanilla"),
+    "alibi_2_0": ModelConfig(4, 24, 24, 128, 2048, 2048, 0.0, "causal", "alibi", alibi_type= "custom"),
+    "alibi_2_1": ModelConfig(2, 24, 24, 128, 2048, 4096, 0.0, "causal", "alibi", alibi_type= "custom"),
```
Comment (timmoon10) [inline tests/pytorch/fused_attn/test_fused_attn.py:401]: Can we get away with smaller configurations? Full correctness testing should be handled in cuDNN and Flash Attention. I propose b=1, h=1, hg=1, d=32, sq={128,1024}, skv={256,2048}.

Comment (cyanguwa) [reply tests/pytorch/fused_attn/test_fused_attn.py:402]: I reduced `b` to 2 and 1, and `sq` to {128, 1024}, `skv` to {256, 2048}. I kept `h`/`hg` and `d` though. I think it's better to test some real-life cases here than not. ALiBi slopes is a different path from some of the other tests.

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -2615,12 +2651,46 @@ def forward(
...
+            fu_core_attention_bias_type = "post_scale_bias"
+            _, fu_core_attention_bias = get_alibi(
+                query_layer.shape[-2], max_seqlen_q, max_seqlen_kv, alibi_slopes=alibi_slopes,
+                bias_dtype=query_layer.dtype)
+            if (fu_core_attention_bias.shape[0] != 1
+                or fu_core_attention_bias.shape[1] != query_layer.shape[-2]):
+                # remove this line when cuDNN adds bwd support for [b, 1, s, s] and [b, h, s, s]
+                use_fused_attention = False
+                # max512 backend will only support [1, h, s, s]
+                os.environ["NVTE_FUSED_ATTN_BACKEND"] = "1"
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:2787]: The max512 backend probably won't support `b != 1` cases so we need to ensure arbitrary_seqlen backend is used.

[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/attention.py:2794]: Ah, I understand now. The `NVTE_FUSED_ATTN_BACKEND` envvar isn't a bool, but is an int corresponding to the specific fused attention backend you want to select. See: https://github.com/NVIDIA/TransformerEngine/blob/e2803b16b0ae94f3167a360063300f9f5179e550/transformer_engine/common/fused_attn/fused_attn.cpp#L166-L172
So we are not enabling the fused attention backend, but rather hinting to use backend 1 (`NVTE_F16_arbitrary_seqlen`).

This isn't documented, so I'll leave this comment unresolved for visibility.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by cyanguwa** [comment]
pipeline 12635563

**Pull request merged**